beacon-chain/rpc/validator/proposer.go

```diff
@@ -194,6 +206,128 @@ func (vs *Server) eth1Data(ctx context.Context, slot uint64) (*ethpb.Eth1Data, e
 	return eth1Data, nil
 }
 
+// eth1DataMajorityVote determines the appropriate eth1data for a block proposal using an extended
+// simple voting algorithm - voting with the majority. The algorithm for this method is as follows:
+//  - Determine if we are in the "head" of the voting period. If yes, use the simple voting algorithm for proposal.
+//  - Determine the timestamp for the start slot for the current eth1 voting period.
+//  - Determine the timestamp for the start slot for the previous eth1 voting period.
+//  - Determine the most recent eth1 block before each timestamp.
+//  - Subtract the current period's eth1block.number by ETH1_FOLLOW_DISTANCE to determine the voting upper bound.
+//  - Subtract the previous period's eth1block.number by ETH1_FOLLOW_DISTANCE to determine the voting lower bound.
+//  - Filter out votes on unknown blocks and blocks which are outside of the range determined by the lower and upper bounds.
```

Comment:
The `unknown blocks` part is something that is not in the linked document, but I added this condition because I feel it is important. Because Eth1Data does not include the block height, it might happen that the block which has the most votes is outside of the range of blocks that we want to vote on. Because of this we cannot safely vote on it, so we need to discard it.

Reply:
:+1: good idea